File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week4/ArrayStack.java
/*
ArrayStack.java
Billy Cussen
12/11/2020
*/
import java.util.Arrays;
public class ArrayStack<T> implements StackInterface<T>{

    private T[] stack;
    private int size;
    private int top;

    public ArrayStack(int size){
        this.size = size;
        this.top = -1;
        this.stack = (T[])new Object[this.size];
    }

    @Override
    public void push(T element){
        if(this.top == this.size-1){
            reallocate();
        } 
        this.top++;
        this.stack[this.top]=element;
    }

    private void reallocate(){
        this.size*=2;
        this.stack = Arrays.copyOf(this.stack,this.size);
    }

    @Override
    public T pop(){
        if(isEmpty()){
            throw new NullPointerException("Can't Pop, the Stack is already Empty!");
        } else{
            T element = this.stack[this.top];
            this.top--;
            return element;
        }
    }

    @Override
    public T peek(){
        if(isEmpty()){
            throw new NullPointerException("Can't Peek, the Stack is Empty!");
        } else{
            return this.stack[top];
        }
    }

    @Override
    public int size(){
        return this.top+1;
    }

    @Override
    public boolean isEmpty(){
        return this.top == -1;
    }

    public String toString(){
        if(isEmpty()){
            return "Can't Traverse Array Stack, its empty!";
        }
        int index = size()-1;
        int position = 0;
        String result = "***ARRAY STACK TRAVERSAL***\n";
        return toStringHelper(this.stack,result,index, position);
    }

    private String toStringHelper(T[] current, String result, int index, int position){
        if(index==-1){
            result+="***END***\n";
            return result;
        }
        result+="Position: "+position+", Element: "+current[index]+"\n";
        return toStringHelper(current,result,--index,++position);
    }

}

File: Python/Data Structures & Algorithms Module - Python/Revision/Stacks/ArrayStack.py
"""
ArrayStack.py
Billy Cussen
07/01/2021
"""
    
class ArrayStack:

    def __init__(self):
        self.top = -1
        self.size = 2
        self.stack = []

    def isEmpty(self):
        return self.top == -1

    def getSize(self):
        if self.isEmpty():
            return 0
        return self.sizeHelper(self.top)

    def sizeHelper(self, index):
        if index == -1:
            return 0
        return 1 + self.sizeHelper(index-1)

    def toString(self):
        if self.isEmpty():
            raise Exception("Can't print, Stack is Empty!")
        result = "***ARRAY STACK TRAVERSAL***\n"
        index = self.getSize()-1
        position = 1
        return self.toStringHelper(self.stack, result, index, position)

    def toStringHelper(self, current, result, index, position):
        if index == -1:
            result+= "***END***\n"
            return result
        result+="Position "+str(position)+", Element: "+str(current[index])+"\n"
        position = position + 1
        index = index - 1
        return self.toStringHelper(current, result, index, position)

    def push(self, element):
        self.top = self.top + 1
        print(str(self.top))
        self.stack.insert(self.top,element)

    def pop(self):
        if self.isEmpty():
            raise Exception("Unable to Pop, Stack is Empty!")
        element = self.stack[self.top]
        self.top = self.top-1
        return element

    def peek(self):
        if self.isEmpty():
            raise Exception("Unable to Peek, Stack is Empty!")
        return self.stack[self.top]
            
